File: backend/app/services/payroll_service.py
"""
Payroll service — computes one month's payroll LIVE, every time it's asked
for: base_salary - deduction_amount + allowance_added + incentive_added -
advance_deducted = net_payable, per employee. Nothing here is persisted or
cached — it's built fresh from attendance_service.get_month_report() (itself
already live) plus this month's allowance/incentive/advance totals, so
entering an allowance, incentive, or advance is reflected the moment you
reopen Payroll, with no separate "generate" step.
"""
import uuid
from sqlalchemy.orm import Session

from . import attendance_service, allowance_service, incentive_service, advance_payment_service


def get_live_payroll(db: Session, hospital_id: uuid.UUID, year: int, month: int) -> dict:
    report = attendance_service.get_month_report(db, hospital_id, year, month)
    allowance_totals = allowance_service.get_month_allowance_totals(db, hospital_id, year, month)
    incentive_totals = incentive_service.get_month_incentive_totals(db, hospital_id, year, month)
    advance_totals = advance_payment_service.get_month_advance_totals(db, hospital_id, year, month)

    items = []
    for emp in report["employees"]:
        # per_day_salary was derived from base_salary / working_days in
        # attendance_service — reconstruct base_salary from it rather than
        # re-querying the user, since working_days is already known here.
        base_salary = round(emp["per_day_salary"] * emp["working_days"], 2)
        user_uuid = uuid.UUID(emp["user_id"])
        allowance_added = float(allowance_totals.get(user_uuid, 0))
        incentive_added = float(incentive_totals.get(user_uuid, 0))
        advance_deducted = float(advance_totals.get(user_uuid, 0))
        net_payable = round(
            base_salary - emp["deduction_amount"] + allowance_added + incentive_added - advance_deducted, 2
        )
        items.append({
            "user_id": emp["user_id"],
            "reference_number": emp["reference_number"],
            "first_name": emp["first_name"],
            "last_name": emp["last_name"],
            "designation": emp["designation"],
            "present_count": emp["present_count"],
            "absent_count": emp["absent_count"],
            "paid_leave_entitlement": emp["paid_leave_entitlement"],
            "working_days": emp["working_days"],
            "base_salary": base_salary,
            "per_day_salary": emp["per_day_salary"],
            "deduction_days": emp["deduction_days"],
            "deduction_amount": emp["deduction_amount"],
            "allowance_added": allowance_added,
            "incentive_added": incentive_added,
            "advance_deducted": advance_deducted,
            "net_payable": net_payable,
        })

    return {
        "year": year,
        "month": month,
        "items": items,
    }

File: frontend/src/services/payrollService.ts
import api from './api';
import type { AllowanceType } from './allowanceService';

export interface PayrollAllowanceLine {
  amount: number;
  reason: string;
  allowance_type: AllowanceType;
}

export interface PayrollIncentiveLine {
  sales_amount: number;
  incentive_percent: number;
  incentive_amount: number;
}

export interface PayrollAdvanceLine {
  amount: number;
  emi_amount: number;
  this_month_deduction: number;
  remaining_after: number;
  reason: string;
}

export interface PayrollItem {
  user_id: string;
  reference_number: string | null;
  first_name: string;
  last_name: string;
  designation: string | null;
  present_count: number;
  absent_count: number;
  paid_leave_entitlement: number;
  working_days: number;
  base_salary: number;
  per_day_salary: number;
  deduction_days: number;
  deduction_amount: number;
  // Sum of this month's "Added to Salary" allowances — already folded into
  // net_payable below. Computed live on every fetch, so entering an
  // allowance/incentive shows up here immediately, no generate step.
  allowance_added: number;
  incentive_added: number;
  // This month's advance-payment EMI deduction(s) — already subtracted
  // from net_payable below.
  advance_deducted: number;
  net_payable: number;
  // Every allowance (both types) logged for this employee this month.
  allowances: PayrollAllowanceLine[];
  incentives: PayrollIncentiveLine[];
  advances: PayrollAdvanceLine[];
}

export interface PayrollRun {
  year: number;
  month: number;
  items: PayrollItem[];
  total_net_payable: number;
}

const payrollService = {
  getPayroll: (year: number, month: number) =>
    api.get<PayrollRun>(`/payroll/${year}/${month}`).then(res => res.data),

  getPayrollPdfUrl: async (year: number, month: number): Promise<string> => {
    const res = await api.get(`/payroll/${year}/${month}/pdf`, { responseType: 'text' });
    return res.data;
  },
};

export default payrollService;
